File: js/singly_linked_list_complete/lib/singly_linked_list.js

function SinglyLinkedListNode(data, nextNode) {
  this.data = data;
  this.nextNode = nextNode ? nextNode : null;
}

exports.SinglyLinkedListNode = SinglyLinkedListNode;

exports.SinglyLinkedList = function(items) {
  this.head = null;
  this.tail = null;
  if (items) {
    for (var i = 0; i < items.length; i++) {
      this.append(items[i]);
    }
  }
};

exports.SinglyLinkedList.prototype.append = function(data) {
  return this.appendNode(new SinglyLinkedListNode(data));
};

exports.SinglyLinkedList.prototype.appendNode = function(node) {
  if (this.tail) {
    this.tail.nextNode = node;
  } else {
    this.head = node;
  }

  var end = node;
  var cur = node.nextNode;
  while (cur) {
    end = cur;
    cur = cur.nextNode;
  }

  this.tail = end;
  return this;
};

exports.SinglyLinkedList.prototype.toString = function() {
  var result = "[";
  var node = this.head;
  if (node) {
    result += JSON.stringify(node.data);
    node = node.nextNode;
    while (node) {
      result += " -> ";
      result += JSON.stringify(node.data);
      node = node.nextNode;
    }
  }
  result += "]";
  return result;
};

exports.SinglyLinkedList.prototype.insert = function(position, data) {
  if (position == 0 && !this.tail) {
    this.append(data);
  } else {
    var node = this.head;
    var prev = null;
    while (node && position > 0) {
      prev = node;
      node = node.nextNode;
      position -= 1;
    }
    if (prev) {
      prev.nextNode = new SinglyLinkedListNode(data);
      prev.nextNode.nextNode = node;
      if (prev == this.tail) {
        this.tail = prev.nextNode;
      }
    } else {
      var old_head = this.head;
      this.head = new SinglyLinkedListNode(data);
      this.head.nextNode = old_head;
    }
  }
  return this;
};

exports.SinglyLinkedList.prototype.delete = function(data) {
  if (!this.head) {
    return this;
  }

  if (this.head.data === data) {
    if (this.head === this.tail) {
      this.head = this.tail = null;
    } else {
      this.head = this.head.nextNode;
    }
  } else {
    var node = this.head;
    while (node.nextNode) {
      if (node.nextNode.data === data) {
        if (!node.nextNode.nextNode) {
          this.tail = node;
        }
        node.nextNode = node.nextNode.nextNode;
      } else {
        node = node.nextNode;
      }
    }
  }
  return this;
};

Object.defineProperty(exports.SinglyLinkedList.prototype, "length", {
  get: function() {
    var len = 0;
    var node = this.head;
    while (node) {
      len++;
      node = node.nextNode;
    }
    return len;
  }
});

exports.SinglyLinkedList.prototype.getItem = function(index) {
  if (index >= 0) {
    var node = this.head;
    while (index > 0 && node)
    {
      node = node.nextNode;
      index--;
    }
    if (node) {
      return node.data;
    }
    return;
  } else {
    var len = this.length;
    if (index * -1 > len) {
      return;
    }
    return this.getItem(len + index);
  }
};

exports.SinglyLinkedList.prototype.forEach = function forEach(callback, thisArg) {
  'use strict';
  var T, k;

  if (this == null) {
    throw new TypeError("this is null or not defined");
  }

  var kValue,
      // 1. Let O be the result of calling ToObject passing the |this| value as the argument.
      O = Object(this);

  // 4. If IsCallable(callback) is false, throw a TypeError exception.
  // See: http://es5.github.com/#x9.11
  if ({}.toString.call(callback) !== "[object Function]") {
    throw new TypeError(callback + " is not a function");
  }

  // 5. If thisArg was supplied, let T be thisArg; else let T be undefined.
  if (arguments.length >= 2) {
    T = thisArg;
  }

  var node = O.head;
  k = 0;

  while (node) {
    kValue = node.data;
    callback.call(T, kValue, k, O);
    node = node.nextNode;
    k++;
  }
};

File: python/singly_linked_list_complete/singly_linked_list.py
class singly_linked_list_node:
  def __init__(self, data = None, next_node = None):
    self.data = data
    self.next_node = next_node

class singly_linked_list:
  def __init__(self, items = None):
    self.head = None
    self.tail = None

    if items:
      for item in items:
        self.append(item)

  def append(self, data):
    """
    >>> l=singly_linked_list()
    >>> print l.append(0)
    [0]
    >>> print l.append(None)
    [0 -> None]
    >>> print l.append("a")
    [0 -> None -> a]
    """
    return self.append_node(singly_linked_list_node(data))
    
  def append_node(self, node):
    """
    >>> l1=singly_linked_list([1,2])
    >>> l2=singly_linked_list([3,4])
    >>> print l1.append_node(l2.head)
    [1 -> 2 -> 3 -> 4]
    >>> print l1.append(5)
    [1 -> 2 -> 3 -> 4 -> 5]
    """
    if self.tail:
      self.tail.next_node = node
    else:
      self.head = self.tail = node

    while self.tail.next_node:
      self.tail = self.tail.next_node

    return self

  def insert(self, position, data):
    """
    >>> l1 = singly_linked_list()
    >>> print l1.insert(0, 'begin')
    [begin]
    >>> print l1.insert(1, 'second')
    [begin -> second]
    >>> print l1.insert(0, 'before begin')
    [before begin -> begin -> second]
    >>> print l1.append('new end')
    [before begin -> begin -> second -> new end]
    """
    if position == 0 and not self.tail:
      self.append(data)
    else:
      node = self.head
      prev = None
      while node and position > 0:
        prev = node
        node = node.next_node
        position -= 1

      if prev:
        prev.next_node = singly_linked_list_node(data)
        prev.next_node.next_node = node
        if prev == self.tail:
          self.tail = prev.next_node
      else:
        old_head = self.head
        self.head = singly_linked_list_node(data)
        self.head.next_node = old_head
    
    return self

  def delete(self, data):
    """
    >>> l=singly_linked_list([10, 20])
    >>> print l.delete(10)
    [20]
    >>> print l.delete(10)
    [20]
    >>> print l.delete(20)
    []
    >>> print l.delete(1)
    []
    >>> l=singly_linked_list([1, 2, 3])
    >>> print l.delete(1)
    [2 -> 3]
    >>> l=singly_linked_list([1, 2, 3])
    >>> print l.delete(2)
    [1 -> 3]
    >>> l=singly_linked_list([1, 2, 3])
    >>> print l.delete(3)
    [1 -> 2]
    >>> print l.delete(2)
    [1]
    >>> print l.delete(1)
    []
    >>> print l.delete(None)
    []
    """
    if not self.head:
      return self

    if self.head.data == data:
      if self.head == self.tail:
        self.head = self.tail = None
      else:
        self.head = self.head.next_node
      return self
			
    node = self.head

    while node.next_node:
      if node.next_node.data == data:
        if node.next_node.next_node == None:
          self.tail = node
        node.next_node = node.next_node.next_node
        return self

      node = node.next_node

    return self

  def __len__(self):
    """
    >>> len(singly_linked_list([]))
    0
    >>> len(singly_linked_list([1]))
    1
    >>> len(singly_linked_list(range(10)))
    10
    """
    length = 0
    node = self.head
    while node:
      length += 1
      node = node.next_node
    return length

  def __getitem__(self, k):
    """
    >>> sll = singly_linked_list([1])
    >>> print sll[0]
    1
    >>> sll = singly_linked_list([1, 2, 3])
    >>> print sll[0]
    1
    >>> print sll[1]
    2
    >>> print sll[2]
    3
    >>> sll = singly_linked_list(["A", "b", "C"])
    >>> sll[0]
    'A'
    >>> sll[-1]
    'C'
    >>> sll[-2]
    'b'
    >>> sll[-3]
    'A'
    >>> sll[-4]
    Traceback (most recent call last):
      ...
    IndexError: list index out of range
    >>> sll[10]
    Traceback (most recent call last):
      ...
    IndexError: list index out of range
    """
    if k >= 0:
      node = self.head
      while node != None and k > 0:
        node = node.next_node
        k -= 1
      if k > 0:
        raise IndexError("list index out of range")
      return node.data
    else:
      length = len(self)
      if k * -1 > length:
        raise IndexError("list index out of range")
      return self[length + k]

  def __iter__(self):
    node = self.head
    while node != None:
      yield node
      node = node.next_node
    return

  def __str__(self):
    """
    >>> print singly_linked_list([0, None, "a"])
    [0 -> None -> a]
    """
    return '[%s]' % ' -> '.join([str(node.data) for node in self])

if __name__ == '__main__':
  import doctest
  result = doctest.testmod()
  if result.failed > 0:
    print "Failed:", result
    import sys
    sys.exit(1)
  else:
    print "Success:", result

